File: EngineeringConcepts/DesignPatterns/1/Logger.cpp
#include "Logger.h"
#include <iostream>
 
Logger::Logger() : logCount(0) {
    std::cout << "Logger instance created at address: " << this << std::endl;
}
 
Logger::~Logger() {
    std::cout << "Logger instance destroyed." << std::endl;
}
 
Logger& Logger::getInstance() {
    static Logger instance;
    return instance;
}
 
void Logger::log(const std::string& message) {
    ++logCount;
    std::cout << "[LOG #" << logCount << "] " << message << std::endl;
}

File: EngineeringConcepts/DesignPatterns/1/Logger.cs
namespace SingletonPatternExample;
 
public sealed class Logger
{
    private static readonly Lazy<Logger> lazyInstance = new Lazy<Logger>(() => new Logger());
 
    private int logCount;
 
    private Logger()
    {
        logCount = 0;
        Console.WriteLine($"Logger instance created with hash code: {GetHashCode()}");
    }
 
    public static Logger Instance => lazyInstance.Value;
 
    public void Log(string message)
    {
        logCount++;
        Console.WriteLine($"[LOG #{logCount}] {message}");
    }
}